db/indexes.go

```diff
@@ -0,0 +1,287 @@
+	indexErrors := make(chan error, len(sgIndexes))
+
+	for _, sgIndex := range sgIndexes {
+		if sgIndex.required {
+			indexesWg.Add(1)
+			go func(index SGIndex) {
+				defer indexesWg.Done()
+				queryStatement := replaceSyncTokens(index.readinessQuery, useXattrs)
+				queryErr := waitForIndex(bucket, index.fullIndexName(useXattrs), queryStatement)
+				if queryErr != nil {
+					indexErrors <- queryErr
+				}
+			}(sgIndex)
+		}
+	}
+
+	indexesWg.Wait()
+	if len(indexErrors) > 0 {
+		err := <-indexErrors
+		close(indexErrors)
+		return err
+	}
+
+	base.Logf("Indexes ready for bucket %s.", bucket.GetName())
+	return nil
+}
+
+// Issues adhoc consistency=request_plus query to determine if specified is ready.  Retries on timeout.
+func waitForIndex(bucket *base.CouchbaseBucketGoCB, indexName string, queryStatement string) error {
+
+	for {
+		_, err := bucket.Query(queryStatement, nil, gocb.RequestPlus, true)
+		// Retry on timeout error, otherwise return
+		if err == nil || err != base.ErrViewTimeoutError {
+			return err
+		} else {
+			base.Logf("Timeout waiting for index %q to be ready for bucket %q - retrying...", indexName, bucket.GetName())
+		}
+	}
+
+}
+
+// Iterates over the index set, removing obsolete indexes:
+//  - indexes based on the inverse value of xattrs being used by the database
+//  - indexes associated with previous versions of the index, for either xattrs=true or xattrs=false
+func removeObsoleteIndexes(bucket base.Bucket, previewOnly bool, useXattrs bool) (removedIndexes []string, err error) {
+
+	gocbBucket, ok := bucket.(*base.CouchbaseBucketGoCB)
+	if !ok {
+		base.Warn("Cannot remove obsolete indexes for non-gocb bucket.")
+		return
+	}
+
+	removedIndexes = make([]string, 0)
+	for _, sgIndex := range sgIndexes {
+		// Remove current version, opposite xattr setting, if present
+		indexName := sgIndex.fullIndexName(!useXattrs)
+		removedIndexes, err = removeObsoleteIndex(gocbBucket, removedIndexes, indexName, previewOnly)
+		if err != nil {
+			return removedIndexes, err
+		}
+
+		// Remove older versions, both xattr and non-xattr, if present
+		for _, prevVersion := range sgIndex.previousVersions {
+			xattrIndexName := sgIndex.indexNameForVersion(prevVersion, true)
+			removedIndexes, err = removeObsoleteIndex(gocbBucket, removedIndexes, xattrIndexName, previewOnly)
+			if err != nil {
+				return removedIndexes, err
+			}
+			nonXattrIndexName := sgIndex.indexNameForVersion(prevVersion, false)
+			removedIndexes, err = removeObsoleteIndex(gocbBucket, removedIndexes, nonXattrIndexName, previewOnly)
+			if err != nil {
+				return removedIndexes, err
+			}
+		}
+	}
+	return removedIndexes, nil
+}
+
+// Removes an obsolete index from the database.  In preview mode, checks for existence of the index only.
+func removeObsoleteIndex(bucket *base.CouchbaseBucketGoCB, removedIndexes []string, indexName string, previewOnly bool) (removedIndexesOut []string, err error) {
```

Comment:
Appending to a slice that gets passed in feels a little awkward to me. I can see its use in `removeObsoleteIndexes` but I think that should be managing its own slice, and have this function return a `removed bool` instead.

```go
func removeObsoleteIndex(bucket *base.CouchbaseBucketGoCB, indexName string, previewOnly bool) (removed bool, err error)
```

Reply:
Good call.